### PR TITLE
BUGFIX: Flush route caches with DocumentUriPathProjection CatchUpHook

### LINKED ISSUE #4377: 9.0 BUG: Accessing a disabled page will throw "SITE NOT FOUND" error
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

After upgrading to the latest Neos 9.0 version and disabling a page in backend the following error will be thrown:

```
TODO: SITE NOT FOUND; should not happen (for address NodeAddress[...]
```

### Expected Behavior

Neos should return a 404 status code

### Steps To Reproduce

_No response_

### Environment

```markdown
- Flow: 9.0
- Neos: 9.0
- PHP: 8.2
```

### Anything else?

_No response_

### LINKED ISSUE #4378: 9.0 BUG: Router cache flush on "documentUriPathChanged" uses wrong tag
When a document uriPath has changed the router cache for the affected nodes/uris need to get flushed. Otherwise the node is still accessable on the old uri.

This current implementation uses the `uriPath` of the DocumentUriPathProjection, which does not contain the dimension prefixes and uri suffix. So the calculated hash/tag does not match the original entry.

https://github.com/neos/neos-development-collection/blob/552e52ebe05b8b3d438d05e7dde51bd692fd2fb5/Neos.Neos/Classes/Package.php#L150

### LINKED ISSUE #4379: 9.0 BUG: Router cache needs get flushed after node aggegate got disabled
When a NodeAggregate gets disabled, we need to flush the cache for the affected nodes. Otherwise the routing will try to show the node, even if they isn't enabled anymore.

See: #4377 

We need a new signal in the DocumentUriPathProjection or catchup hook to handle the cache flush properly. 
See also: https://github.com/neos/neos-development-collection/issues/4378

Actually we should also check if it is needed after enabling a node too.

### PR DESCRIPTION
**Pull request by dlubitz**
Fixes #4379 #4378

### CODE REVIEW

File: Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\ContentRepository\Core\DimensionSpace\DimensionSpacePoint;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\Factory\ContentRepositoryId;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\ContentRepository\Core\Feature\NodeDisabling\Event\NodeAggregateWasDisabled;
+use Neos\Neos\Routing\Cache\RouteCacheFlusher;
+use Neos\Flow\Mvc\Routing\RouterCachingService;
+
+final class DocumentUriPathProjectionHookForRouteCacheFlushing implements CatchUpHookInterface
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php]: Just a nitpick, but this is a veery long class name, what about just `RouterCacheHook` and a short class doc comment?

File: Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+{
+    /**
+     * Runtime cache to collect tags until they can get flushed.
+     * @var array<string, string>
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php]: could you document the format of the array? is this really always string key and value?

File: Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private function collectTagsToFlush(DocumentNodeInfo $node): void
+    {
+        array_push($this->tagsToFlush, ...$node->getRouteTags()->getTags());
+    }
+
+    private function flushAllCollectedTags(): void
+    {
+        $this->routerCachingService->flushCachesByTags($this->tagsToFlush);
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php]: This should probably have an empty-check
```suggestion
        if ($this->tagsToFlush === []) {
            return;
        }
        $this->routerCachingService->flushCachesByTags($this->tagsToFlush);
```

File: Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private function getState(): DocumentUriPathFinder
+    {
+        return $this->contentRepository->projectionState(DocumentUriPathFinder::class);
+    }
+
+    private function isLiveContentStream(ContentStreamId $contentStreamId): bool
+    {
+        return $contentStreamId->equals($this->getState()->getLiveContentStreamId());
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php]: we could move that method to the `DocumentUriPathFinder` IMO, but that's just a mini tweak ofc

**Review by bwaidelich — COMMENTED** [review]
Nice one, thank you!
I added a couple of comments, most of which you are free to ignore of course :)

Actually I think this is more of a *bugfix* than a *feature* really (not making it any less nice or important of course). And I have the urge to cover this with tests – but that's probably not so easy..

File: Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php (hunk 1 above)
Comment (dlubitz) [reply Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php]: Sure, I was inspired by `GraphProjectorCatchUpHookForCacheFlushing` but I don't insist on anything.

**Comment by dlubitz** [comment]
WDYT about removing the signal "documentUriPathChanged" in the DocumentUriPathProjection?
https://github.com/neos/neos-development-collection/blob/cc4a0a68657b7f72e95360cd4e99a1e93b21ab67/Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php#L622

**Comment by bwaidelich** [comment]
> WDYT about removing the signal "documentUriPathChanged" in the DocumentUriPathProjection?

yes, absolutely. Signal/Slot won't really make sense with everything that is related to a single CR instance

**Comment by dlubitz** [comment]
TODO: 
* Remove tryGetNode and replace with method 
* Merge 9.0 to fix pipeline

**Review by ahaeslich — APPROVED** [review]
thx 💖 also tested it in our project

**Review by kitsunet — APPROVED** [review]

**Pull request merged**